### ISSUE TITLE
Sector set storage

### ISSUE THREAD

**Issue by whyrusleeping** [filecoin-project/specs#116]
We need a data structure to hold sectors on chain, inside the miner actor. Currently, we just have them all in a big map, which won’t scale well (every change to it creates a new object, meaning we would rewrite the entire thing with every change).

This issue is a place to brainstorm solutions.

#### Requirements
* Ordered Set
	* no values, just keys (TODO: is this true? Is it just storing CommRs?)
	* ~~‘Ordered’ requirement isn’t as strict as it sounds, we really just need to be able to reference elements within a given set by their index within the set. It is okay if indexes of items change with insertions and deletions.~~
* Efficient additions and random removals
* ~O(log(n)) changed objects per mutation
* Efficient lookup by key and index

#### Nice to haves
* Efficient batch remove by index range

### Does the HAMT work?
We could potentially use the HAMT code, but we would need to adapt it to have efficient index lookups. To do this, we could annotate links to children with a count of how many values exist in and under that child, allowing us to know at each node which child the thing we are looking up is in.

**Comment by porcuquine** [comment]
FWIW, replicaID currently includes the sectorID.

**Comment by whyrusleeping** [comment]
@porcuquine I just read your comments, youre basically saying "if miners can find hash function collisions, they can pull off an attack". To which I respond: If you can generate collisions, the hash function is broken.

If we are assuming that the hash function will not be broken anyways, then i see no problem in allowing full miner selection over the ID.

In addition, I believe that it is a hard requirement that miners have full selection over the sector IDs (I could bring this down to full selection within some range) in order for the protocol to work scalably.

**Comment by porcuquine** [comment]
@whyrusleeping I actually agree with you and concluded the same in my edit.

This doesn't apply to the risks of grinding to generate a fake Seal proof. Although this doesn't add to the risk during a generation attack, allowing an infinite sampling of replica IDs does (I believe) reduce attempts to generate a single fake PoRep for a fixed `commD` to being as easy (or hard) as for any other generation attack.

While this doesn't affect security of the power table, it does reduce the security of storage market redundancy to 'only' on what is also provided against generation attacks. Of course, we assume this is sufficient, but I do think this is a reduction — so we should at least be clear about it.

**Comment by whyrusleeping** [comment]
Additional interesting design constraint to consider: **Do we ever allow sector ID reuse?** If we do not, then we have to track which IDs have been used, in order to disallow their use in the future.

If we do allow for ID reuse, then I think a fairly straightforward potential solution would be to use the persistent vector, but modify it to allow for sparse regions. Such a data structure should satisfy all the design constraints. I'll look into this more soon.

**Comment by whyrusleeping** [comment]
I started to implement a sparse persistent vector here, and realized that many pieces are very similar to the HAMT. Particularly, when making it sparse, you end up wanting to use the same bitfield compression prefix count optimizations that you use for HAMTs. so now i'm thinking, maybe we use the HAMT still, but swap out the functionality that turns a key into an index at each height.

**Comment by whyrusleeping** [comment]
Some notes i sketched on this idea:
![data structures -5](https://user-images.githubusercontent.com/1243164/53671826-eb96b800-3c34-11e9-8f5c-df06dea3ef6f.jpg)
![data structures -6](https://user-images.githubusercontent.com/1243164/53671837-f94c3d80-3c34-11e9-8819-19a290fdabcc.jpg)

**Comment by whyrusleeping** [comment]
@porcuquine thoughts?

**Comment by anorth** [comment]
For context, could you please clarify whether this is a data structure that will appear literally in the on-chain bytes (like the encoding of messages), or is it a private implementation detail on which different nodes need not necessarily agree. In the latter case, the issue might belong in an implementation repo rather than `specs`.

**Comment by nicola** [comment]
note that the SNARK must be able to signal what are the failing sectors, so either this is snark friendly or the verifier will need to compute the inputs to verify the snark later on (in other words: either "compressing" the sectors is cheap or "decompressing" the sector must be cheap and done out of the snark)

**Comment by porcuquine** [comment]
@whyrusleeping After a few days of kicking this around, I don't see any big problems with your AMT and haven't come up with any incremental improvements. Either might come as more time passes, but in accordance with what we discussed earlier, you can take this as a first vote of confidence toward solidifying this design.

**Comment by whyrusleeping** [comment]
@anorth the first sentence of the issue is: "We need a data structure to hold sectors on chain, inside the miner actor." 

Should I clarify in some other way?

**Comment by whyrusleeping** [comment]
The question of sector ID reuse has come up again in the context of payments. Payments to miners need to be contingent on a given sector, this can be via a reference to a sector ID, but if miners can reuse sector IDs, then they could drop the data the payment is for and then add new data with the same sector ID.

This sucks, but preventing sector ID reuse while maintaining the miners free choice over sector IDs to use is very difficult.

An alternative structure for the payments that could work here is to reference the sector by its commD hash. This isnt easy to look up, as it requires a scan of the sector set to confirm its there, but we can short circuit that by having the miner provide the sector ID that has the given commD, the actor code can then verify that it matches.

tldr; I think sector ID reuse is fine, we just have to be aware of it

**Comment by whyrusleeping** [comment]
Note: depends on PoSt faults requirements

**Comment by whyrusleeping** [comment]
Note for @nicola : 

Check whether we need to implement faults in PoSt, or if we can get away with having miners who encounter faults post their proof on chain immediately.

**Comment by dignifiedquire** [comment]
> Check whether we need to implement faults in PoSt, or if we can get away with having miners who encounter faults post their proof on chain immediately.

This would still need some form of implementation in PoSt, as an "aborted PoSt" still needs different proof verification, compared to a completed one. Or am I missing something here?

**Comment by whyrusleeping** [comment]
@dignifiedquire I think this would require arbitrary length PoSts

**Comment by dignifiedquire** [comment]
@whyrusleeping some questions that came up when trying to actually implement the amt

1. Should we do the CHAMP optimization of splitting things into two different arrays (one for references, one for values)
  - this depends on how important in order traversal is, as it gets very inefficient trying to do in order traversal when doing the regular version
  - as far as I can tell this optimization is not part of go-hamt-ipld
2. Should we do the bitset optimizations to avoid storing very large arrays?
3. What base encoding should we use to represent the keys/indexes when translating them?


> Efficient lookup by key and index

What is the difference between key and index for you? I was assuming we use the index as key.

**Comment by whyrusleeping** [comment]
> Should we do the CHAMP optimization of splitting things into two different arrays (one for references, one for values)

Either way really, I think if we use a CBOR map as the top level thing, then it doesnt cost us anything extra to have one array for each (empty arrays in maps are free because you just leave the field out).

I don't see how it affects order traversal, mind elaborating?

> Should we do the bitset optimizations to avoid storing very large arrays?

That seems to make sense to me, but it depends on the expected case. If we're going to have *mostly* full nodes, then its just overhead. So for now, maybe we don't do that optimization?

> What base encoding should we use to represent the keys/indexes when translating them?

Translating? Inside the data structure, the keys should just be bytes. and yeah, we don't really have keys, just indexes. Which means we don't even need to store them, as you can infer an index by the nodes location within the structure.

**Comment by dignifiedquire** [comment]
> I don't see how it affects order traversal, mind elaborating?

I am referring to the first few paragraphs here: https://blog.acolyer.org/2015/11/27/hamt/ under "A new champ".

**Comment by dignifiedquire** [comment]
>  and yeah, we don't really have keys, just indexes.

👌 


> Which means we don't even need to store them, as you can infer an index by the nodes location within the structure.

👌 that was what I thought as well

**Comment by dignifiedquire** [comment]
> Translating? Inside the data structure, the keys should just be bytes.

The issue that I have is that your diagram above makes it look like we try to get things into the tree such that a depth first traversal would lead to an in order traversal of the nodes (using base10). This is not the case if we use a base2 representation, as far as I can tell.

**Comment by dignifiedquire** [comment]
Rereading your diagram, the ordering is definitely not there, even with base10, so ignore that part.

**Comment by whyrusleeping** [comment]
@dignifiedquire you should be able to have trivial in-order traversal of the indexes, thats most of the point. Unless i'm missing something.

The bucketing algorithm should be roughly, at each layer, `child = index / width` and then when you recurse (if you recurse) your index within the child is `index_in_child = index % width`.

**Comment by ZenGround0** [comment]
Hey @whyrusleeping @dignifiedquire I have some basic questions.  Bear with me while I get up to speed here.

Is the proposal to use the AMT datastructure from "Fast And Space Efficient Trie Searches" with the keys being the sectorID indexes?  Are you proposing anything different when "extending the range" than an AMT would do? If so it's not obvious from the picture and I would love some clarification.   I would also appreciate a definition of the "stacking process" if it differs from a canonical AMT.

Also, is this inspired by section 5 of the HAMT paper "SORTED ORDER AMT"?  If not it seems like there are a few promising ideas closely related to our objective that might be worth thinking over.

**Comment by dignifiedquire** [comment]
- I have not heard of "SORTED ORDER AMT", so from my side there is nothing that was taken from there (yet). 
- Yes the idea is to use the indexes as keys, instead of hashing them 
- we should be pre expanding the tree so that all insertions are in their canonical place, such that we can determine the index be reading the tree, without having to actually store it
- I am not sure what you are referring to with "stacking process"

**Comment by dignifiedquire** [comment]
AMT the name comes from "HAMT" -> no hashing, leaves us with "AMT"

**Comment by ZenGround0** [comment]
"stacking process" is just referencing the phrase "to insert any larger value repeat the stacking process" in the diagram above.

@whyrusleeping why doesn't the diagram above have many intermediate tables like this:
![Screen Shot 2019-06-04 at 7 17 21 PM](https://user-images.githubusercontent.com/5515260/58978711-63c3cb00-879a-11e9-9010-74615ae142b4.png)

It seems like we need to traverse the trie starting with the most significant symbols of the index and working down to the least significant symbols of the index to get the correct ordering.  But in that case wouldn't `0` and `2` have O(log(max sector set size)) collisions during insertion?

**Comment by whyrusleeping** [comment]
@ZenGround0 thats part of the point here, the structure is only deep as the log of its biggest element. It expands as you insert increasingly larger elements into it (and would shrink if you removed them)

**Comment by ZenGround0** [comment]
Ok I'm closing in.  My remaining questions are:
1) What node width do we want?  go-ipld-hamt currently has an implied width of 256 pointers per node.  Should we go with that here too?
2) As an optimization go-ipld-hamt uses linear probing over `W=3` nodes by comparing the search key against the stored key in the `KVs` array.  I think we cannot do this because we want to avoid storing the key.  Is this right?
3) Furthermore I don't think we can even do the optimization in why's diagram where we only allocate a new node for 33 *after* 37 comes along.  I think we need to allocate a new node when we first see 33.  Again this is because we are planning to not store the key.  Is this right?
4) Above @dignifiedquire  asks "Should we do the bitset optimizations to avoid storing very large arrays?" and @whyrusleeping  responds "So for now, maybe we don't do that optimization?".  I am taking that to mean that we allocate the entire node width (say 256 Pointers) and not include a bitfield in the node at all.  Does this sound good?

**Comment by whyrusleeping** [comment]
I implemented an AMT here: https://github.com/filecoin-project/go-amt-ipld
